### PR TITLE
Check SqlPrinter round trip when analyzing or planning statements

### PR DESCRIPTION
**Pull request by mxm**
This enables two checks for all unit tests which use the `SQLExecutor`:

1) Statements which are analyzed, are analyzed and printed twice. The printout
of both statements has to match.

2) Statements which are planned, are printed and planned again to check if both
plans match.

This provides additional coverage in addition to the `SqlPrinter`.

**NOTE**: There are failing tests (103) which need to be fixed before we can merge this.

### CODE REVIEW

**Comment by chaudum** [comment]
@mxm what are the issues of the failing tests related to?

**Comment by mxm** [comment]
@chaudum There are some more issues to address for the SqlPrinter.

- MATCH queries (see additional pushed commit)
- subscript notation
- cast/try cast functions
- distinct queries
- collect_set
- ANY queries

There could be more. Will ping you when all the tests are fixed.

**Comment by mxm** [comment]
@chaudum @mfussenegger There are still two test types failing: `SemiJoinsTest.testWriteWithMultipleInClauses` and `WhereClauseAnalyzerTest`. See https://jenkins.crate.io/job/CrateDB/job/crate_on_pr/427/

Should be possible to fix the semi joins tests by refining our cast strategy for subqueries. Tried but failed so far.

As for the`WhereClauseAnalyzerTest`, @chaudum disabled the rountrip in the tests but I think we have to come up with a way to only apply the where clause optimization once instead. The optimization is currently applied again in the round trip which alters the statement and breaks the roundtrip check. I've reverted the commit so we can see the failing tests.

File: sql/src/main/java/io/crate/expression/symbol/SelectSymbol.java [hunk 1]
```diff
@@ -73,6 +77,20 @@ public boolean canBeCasted() {
...
 
+    @Override
+    public Symbol cast(DataType newDataType, boolean tryCast) {
+        if (DataTypes.isCollectionType(newDataType)) {
+            List<Symbol> castSymbols = new ArrayList<>(relation.outputs().size());
+            DataType innerType = ((CollectionType) newDataType).innerType();
+            for (Symbol symbol : relation.outputs()) {
+                castSymbols.add(symbol.cast(innerType, tryCast));
+            }
+            relation.querySpec().outputs(castSymbols);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/symbol/SelectSymbol.java]: This is problematic because the types of the fields won't match the outputs. Why is this necessary now?

File: sql/src/main/java/io/crate/expression/symbol/format/SymbolPrinter.java [hunk 2]
```diff
@@ -182,9 +212,92 @@ private String anyOperatorName(String functionName) {
...
+        private void printMatchPredicate(Function matchPredicate, SymbolPrinterContext context) {
+            StringBuilder sb = context.builder;
+            // Due to the match predicate being a hack and not an actual function, we have to manually print it here
+            sb.append("MATCH(");
+            List<Symbol> arguments = matchPredicate.arguments();
+            // first argument (cols)
+            Map columnBootMap = (Map) ((Literal) arguments.get(0)).value();
+            sb.append("(");
+            Iterator<Map.Entry<String, Object>> entryIterator = columnBootMap.entrySet().iterator();
+            while (entryIterator.hasNext()) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/symbol/format/SymbolPrinter.java]: would it make sense to split this method up a bit?

File: sql/src/main/java/io/crate/expression/symbol/format/SymbolPrinter.java [hunk 3]
```diff
@@ -182,9 +212,92 @@ private String anyOperatorName(String functionName) {
...
+                context.builder.append("[");
+                process(arguments.get(1), context);
+                context.builder.append("]");
+            }
+        }
+
         @Override
         public Void visitReference(Reference symbol, SymbolPrinterContext context) {
-            if (context.isFullQualified()) {
+            if (context.isFullQualified() && !symbol.ident().tableIdent().equals(UnnestFunction.TABLE_IDENT)) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/symbol/format/SymbolPrinter.java]: why do we have to handle unnest special but not other table functions?

File: sql-parser/src/main/antlr/SqlBase.g4 [hunk 4]
```diff
@@ -527,6 +528,10 @@ setTypeDefinition
     : SET '(' dataType ')'
     ;
 
+tableTypeDefinition
+    : TABLE '(' dataType ')'
```
Comment (mfussenegger) [inline sql-parser/src/main/antlr/SqlBase.g4]: How does the SQL look like where we need this to parse it?

Comment (mxm) [reply sql-parser/src/main/antlr/SqlBase.g4]: Would look like this:

```sql
Select * from where col = ANY(cast((Select col2 from t) as table(col_type)))
```

Whenever a cast is necessary, i.e. we're not selecting literals which can be converted without a cast.

File: sql/src/main/java/io/crate/expression/symbol/format/SymbolPrinter.java (hunk 2 above)
Comment (mxm) [reply sql/src/main/java/io/crate/expression/symbol/format/SymbolPrinter.java]: It's rather lengthy due to the complexity that comes with printing the MATCH clause. There are some comments to mark the different parts. On the one hand, breaking it up would make it more readable but on the other hand we would end up with parts which are not used independently of each other.

File: sql/src/main/java/io/crate/expression/symbol/SelectSymbol.java (hunk 1 above)
Comment (mxm) [reply sql/src/main/java/io/crate/expression/symbol/SelectSymbol.java]: The idea here was to ad-hoc replace the literals inside the SelectSymbol. Maybe not a good idea, we should go through the regular normalization process instead.

File: sql-parser/src/main/antlr/SqlBase.g4 (hunk 4 above)
Comment (mfussenegger) [reply sql-parser/src/main/antlr/SqlBase.g4]: I'd like to avoid introducing a user-facing `table` type. Can we handle this in a different way?

File: sql/src/main/java/io/crate/expression/symbol/format/SymbolPrinter.java (hunk 2 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/expression/symbol/format/SymbolPrinter.java]: Maybe introduce a dedicated `MatchPrinter` ?

File: sql/src/main/java/io/crate/expression/symbol/format/SymbolPrinter.java (hunk 3 above)
Comment (mxm) [reply sql/src/main/java/io/crate/expression/symbol/format/SymbolPrinter.java]: Hmm good point. This should generally apply to table functions.

File: sql-parser/src/main/java/io/crate/sql/Identifiers.java [hunk 5]
```diff
@@ -53,7 +53,7 @@ public static String quoteIfNeeded(String identifier) {
     private static boolean areQuotesRequired(String identifier) {
         for (int i = 0; i < identifier.length(); i++) {
             int cp = identifier.codePointAt(i);
-            if (cp == '"' || Character.isUpperCase(cp)) {
+            if (cp == '"' || cp == '[' || cp == ']' || cp == '(' || cp == ')' || Character.isUpperCase(cp)) {
```
[the remark this answers is not in the document]
Comment (mxm) [reply sql-parser/src/main/java/io/crate/sql/Identifiers.java]: Yes, but that is handled in the next step where we create the identifier from the parser.

[the remark this answers is not in the document]
Comment (chaudum) [reply sql-parser/src/main/java/io/crate/sql/Identifiers.java]: Ah ok, got it!

File: sql-parser/src/main/antlr/SqlBase.g4 (hunk 4 above)
Comment (mxm) [reply sql-parser/src/main/antlr/SqlBase.g4]: I'm afraid, I don't see how it can be avoided since TABLE is the real type which the subquery returns. If we convert from int->long or the other way, we need to apply a table cast.

Perhaps just get rid of the table type and use array? There is no difference currently between array and table, the table is represented as an array.

File: sql/src/main/java/io/crate/expression/symbol/SelectSymbol.java (hunk 1 above)
Comment (mxm) [reply sql/src/main/java/io/crate/expression/symbol/SelectSymbol.java]: Reworked this in https://github.com/crate/crate/pull/7195/commits/91789cf20b1985e31ec7ca2e88ab34dc6716983f

File: sql/src/main/java/io/crate/expression/symbol/format/SymbolPrinter.java (hunk 2 above)
Comment (mxm) [reply sql/src/main/java/io/crate/expression/symbol/format/SymbolPrinter.java]: Could do that but there is already a dedicated method `printMatchPredicate`. I don't see the added benefit of having an extra class.

**Comment by mxm** [comment]
@chaudum @mfussenegger This is ready for a final review.

**Comment by mxm** [comment]
>java.nio.file.FileSystemException: /var/lib/jenkins/workspace/CrateDB/crate_on_pr: No space left on device

**Comment by mxm** [comment]
retest this please

File: sql/src/main/java/io/crate/analyze/SQLPrinter.java [hunk 6]
```diff
@@ -113,25 +121,29 @@ public Void visitMultiSourceSelect(MultiSourceSelect mss, StringBuilder sb) {
...
-            addFrom(sb, relation);
-            clauseAndQuery(sb, "WHERE", relation.where());
+            WhereClause whereClause = addFrom(sb, relation);
+            addQueryClauses(relation, whereClause, sb);
+            return null;
+        }
+
+        private void addQueryClauses(QueriedRelation relation, WhereClause whereClause, StringBuilder sb) {
+            // We need to convert any References back to keep any alias information
+            QuerySpec querySpec = relation.querySpec().copyAndReplace(
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/SQLPrinter.java]: Would it make sense to apply the transformation on the components individually so that we can keep referring to the properties on the relation? We want to get rid of the QuerySpec in its current form and these changes here will make that harder.

File: sql/src/main/java/io/crate/analyze/SQLPrinter.java [hunk 7]
```diff
@@ -265,63 +291,107 @@ private void addFrom(StringBuilder sb, QueriedRelation relation) {
...
                 }
-                sb.append(" ON ");
-                sb.append(printSymbol(currentJoinPair.condition()));
             }
-
+            return finalWhereClause;
         }
 
-        private static void printRelationName(AnalyzedRelation relation, StringBuilder sb) {
+        private WhereClause printRelationName(AnalyzedRelation relation, StringBuilder sb) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/SQLPrinter.java]: That `printRelationName` returns a `WhereClause` is rather strange. Can we change the method name or otherwise re-structure this?

File: sql/src/main/java/io/crate/expression/symbol/format/SymbolPrinter.java (hunk 2 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/expression/symbol/format/SymbolPrinter.java]: Would help in being able to introduce some private methods to split up the code here without cluttering the SymbolPrinter

File: sql/src/main/java/io/crate/expression/symbol/format/SymbolPrinter.java [hunk 8]
```diff
@@ -274,6 +392,20 @@ private void printArgs(List<Symbol> args, SymbolPrinterContext context) {
                 }
             }
         }
+
+        private static boolean isTableFunctionReference(Reference reference) {
+            RelationName relationName = reference.ident().tableIdent();
+            return relationName.equals(UnnestFunction.TABLE_IDENT) || relationName.equals(EmptyRowTableFunction.TABLE_IDENT);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/symbol/format/SymbolPrinter.java]: This is bound to break as soon as we add other table functions. I think we need to come up with a different way to detect if a table function is involved. Or change the structure somehow.

**Comment by mxm** [comment]
@mfussenegger Addressed your comments in fixups.

File: sql/src/main/java/io/crate/analyze/MultiSourceSelect.java [hunk 9]
```diff
@@ -93,12 +93,22 @@ public static MultiSourceSelect createWithPushDown(RelationNormalizer relationNo
...
+        Map<QualifiedName, AnalyzedRelation> newSources = new LinkedHashMap<>();
+        for (Map.Entry<QualifiedName, AnalyzedRelation> source : mss.sources.entrySet()) {
+            AnalyzedRelation analyzedRelation = source.getValue();
+            newSources.put(analyzedRelation.getQualifiedName(), analyzedRelation);
+        }
+
+        Function<QualifiedName, QualifiedName> sourceReplaceFun = qName -> mss.sources().get(qName).getQualifiedName();
         for (JoinPair joinPair : mss.joinPairs) {
             joinPair.replaceCondition(convertFieldInSymbolsToNewRelations);
+            joinPair.replaceSourceNames(sourceReplaceFun);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/MultiSourceSelect.java]: I wonder if we should simply create new `joinPairs` given that we're creating a new `MultiSourceSelect` anyway?

File: sql/src/main/java/io/crate/expression/symbol/format/MatchPrinter.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+                                    SymbolPrinter.SymbolPrintVisitor symbolPrintVisitor) {
+        StringBuilder sb = context.builder;
+        // Due to the match predicate being a hack and not an actual function, we have to manually print it here
+        sb.append("MATCH(");
+        List<Symbol> arguments = matchPredicate.arguments();
+        // first argument (cols)
+        Map columnBootMap = (Map) ((Literal) arguments.get(0)).value();
+        sb.append("(");
+        Iterator<Map.Entry<String, Object>> entryIterator = columnBootMap.entrySet().iterator();
+        while (entryIterator.hasNext()) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/symbol/format/MatchPrinter.java]: now that it is in its own class splitting this up a bit would be nice, like

```
printColumns(..)
printKeyword(..)
printUsing(..)
printWith(..)
```
